specs/capri/Multi-attach.md

```diff
@@ -0,0 +1,187 @@
+	TotalCapacity int64 `json:"totalCapacity,omitempty"`
+	FreeCapacity int64 `json:"freeCapacity,omitempty"`
+	StorageType string `json:"storageType,omitempty"`
+	Extras StoragePoolExtraSpec `json:"extras,omitempty"`
+	ReplicationType string `json:"replicationType,omitempty"`
+	ReplicationDriverName string `json:"replicationDriverName,omitempty"`
+
+	// MultiAttach
+	// If true, this volume can attach to more than one instance. Default will be multiAttach:False
+	MultiAttach bool `json:"multiAttach,omitempty"`
+}
+```
+
+
+2. The volume attachment struct adds AttachMode, read-only (‘ro’) or read-and-write (‘rw’), default is ‘rw’.
+
+```
+type VolumeAttachmentSpec struct {
+	*BaseModel
+	TenantId string `json:"tenantId,omitempty"`
+	UserId string `json:"userId,omitempty"`
+	VolumeId string `json:"volumeId,omitempty"`
+	Mountpoint string `json:"mountpoint,omitempty"`
+	Status string `json:"status,omitempty"`
+	Metadata map[string]string `json:"metadata,omitempty"`
+	HostInfo `json:"hostInfo,omitempty"`
+	ConnectionInfo `json:"connectionInfo,omitempty"`
+	AccessProtocol string `json:"accessProtocol,omitempty"`
+	AttachMode string `json:"attachMode,omitempty"`
+}
+```
+
+3. The volume struct adds MultiAttach.
+
+```
+type VolumeSpec struct {
+	*BaseModel
+	TenantId string `json:"tenantId,omitempty"`
+	UserId string `json:"userId,omitempty"`
+	Name string `json:"name,omitempty"`
+	Description string `json:"description,omitempty"`
+	GroupId string `json:"groupId,omitempty"`
+	Size int64 `json:"size,omitempty"`
+	AvailabilityZone string `json:"availabilityZone,omitempty"`
+	Status string `json:"status,omitempty"`
+	PoolId string `json:"poolId,omitempty"`
+	ProfileId string `json:"profileId,omitempty"`
+	Metadata map[string]string `json:"metadata,omitempty"`
+	SnapshotId string `json:"snapshotId,omitempty"`
+	SnapshotFromCloud bool `json:"snapshotFromCloud,omitempty"`
+	ReplicationId string `json:"replicationId,omitempty"`
+	ReplicationDriverData map[string]string `json:"replicationDriverData,omitempty"`
+	AttachStatus string
+	MultiAttach bool `json:"multiAttach"`
+}
+```
+
+### REST API impact
+
+N/A
+
+### Security impact
+
+N/A
+
+### Other end user impact
+
+N/A
+
+### Performance impact
+
+User first needs to use the following command to create a multi-attach related profile and specify it in volume creation.
+
+```
+osdsctl profile create '{"name": "multi-attach", "description": "multi attach profile","customProperties":{"multiAttach":"<is> true"}}'
+```
+
+### Other deployer impact
+
+N/A
+
+### Developer impact
+
+Drivers will need to add a capabilities field "multiAttach: True/False" to their pool configuration files, and do any special handling on their end for connecting/disconnecting volumes in this category.
+
+## Use Cases
+
+1. There are products such as Oracle RAC that can leverage this feature. Oracle RAC uses H/A database via a shared block device.
+2. Passive stand-by servers/volumes.
```

Comment:
Maybe we can add that: multi-attach may be required active/active or active/passive scenario.

Reply:
Sure.